### PR TITLE
ARROW-18247: [JS] fix: RangeError crash in Vector.toArray()

### PR DESCRIPTION
**Pull request by sarfata**
This fixes a bug which happens when a Vector has been created with multiple data blobs and at least one of them has been padded.

See https://observablehq.com/d/14488c116b338560 for a reproduction of the error and more details.

### CODE REVIEW

**Review by domoritz — APPROVED** [review]
Thank you for the pull request. Please add a unit test so we make sure this issue doesn't surface again.

**Review by trxcllnt — APPROVED** [review]
Nice catch, thanks for the fix @sarfata!

**Comment by sarfata** [comment]
@domoritz @trxcllnt Thanks for your reviews.

I have added a unit test for my scenario. I also realized that my fix did not work when stride is not 1 so I have added a test (and fix) for that too.

I believe one of you will need to approve running the test workflow again before merging.

File: js/test/unit/vector/vector-tests.ts [hunk 1]
```diff
@@ -211,6 +211,34 @@ describe(`ListVector`, () => {
     });
 });
 
+describe(`toArray()`, () => {
+    test(`when some data blobs have been padded`, () => {
+        const d1 = vectorFromArray([...Array(16).keys()]);
+        const d2 = vectorFromArray([...Array(10).keys()]);
```
Comment (trxcllnt) [inline js/test/unit/vector/vector-tests.ts:217]:
```suggestion
        const d1 = vectorFromArray([...new Array(16).keys()]);
        const d2 = vectorFromArray([...new Array(10).keys()]);
```

File: js/test/unit/vector/vector-tests.ts [hunk 2]
```diff
@@ -211,6 +211,34 @@ describe(`ListVector`, () => {
...
+
+        // Padding has been added
+        expect(d2.length < d2.data.length)
```
Comment (trxcllnt) [inline js/test/unit/vector/vector-tests.ts:220]:
```suggestion
        expect(d2.length < d2.data.length);
```

**Comment by domoritz** [comment]
Looks like a test is failing. 

Screenshot 2022-12-09 at 17 21 24

**Comment by sarfata** [comment]
@domoritz fixed. 

🤦 I did some last minute cleanup and added this test to show exactly the key condition for this bug but ended up with a failing test. Sorry about this.

**Pull request merged**